csrc/id_model/predicate_indexing.cpp

```diff
@@ -53,4 +53,104 @@ std::vector<IterDomain*> getPredicateDomains(
   return predicate_domains;
 }
 
+std::unordered_map<Val*, Val*> getPredicateIndexReplacementMap(
+    TensorView* tv,
+    const std::vector<ForLoop*>& for_loops,
+    const std::unordered_map<ValGroup, Val*>& index_map,
+    const ValGraph& traversal_graph,
+    const IdModel& id_model,
+    bool is_start_predicate,
+    ForLoop* unswitched_loop) {
+  std::unordered_map<Val*, Val*> replacement_map;
+
+  // For an iter domain of index i, it is valid to use N-1 instead of
+  // i, where N is the extent of the iter domain if either of the
+  // following conditions is satisfied:
+  //
+  // - Vectorized
+  // - predicateAtEnd returns true
+  // - Within an unswitch/unroll loop
+  //
+  // Use N-1 instead of i but not when it's thread paralellized so
+  // that each thread or block can take different paths. This may not
+  // be optimal for TID, though, as it might result in thread
+  // divergence.
+  //
+  // Also in the case of vectorization, instead of N-1, it's also
+  // valid to use 0 since the splits involved to create the iter
+  // domain are all guaranteed to be divisible.
+  auto predicate_at_end =
+      [&](ForLoop* fl, IterDomain* loop_id, bool within_unswitch) -> Val* {
+    // Don't replace thread indices even when unswitched
+    if (!fl->iter_domain()->isThread() &&
+        (fl->iter_domain()->getParallelType() == ParallelType::Vectorize ||
+         within_unswitch || lower_utils::predicateAtEnd(fl))) {
+      return is_start_predicate
+          ? fl->fusion()->zeroVal()
+          : SimplifyingIrBuilder::subExpr(
+                fl->simplifiedStop(), fl->fusion()->oneVal());
+    } else {
+      return nullptr;
+    }
+  };
+
+  // Inspect the for-loops from outer to inner and keep track of
+  // unswitching since it affects all inner loops
+  bool within_unswitch = false;
+  for (const auto fl : for_loops) {
```

Comment:
Pretty much the same logic as `getPredicateInitialIndexParameters` in analysis/index_compute.cpp